javalib/src/main/scala/java/util/regex/Pattern.scala

```diff
@@ -19,243 +22,116 @@ object Pattern {
-    CompiledPatternStore.withRE2Regex(regex, flags)(_ => ())
+    def notSupported(flag: Int, flagName: String): Unit = {
+      if ((flags & flag) == flag) {
+        throw new UnsupportedOperationException(
+          s"regex flag $flagName is not supported.")
+      }
+    }
 
-    new Pattern(
-      _pattern = regex,
-      _flags = flags
+    val unsupportedOptions = Array[(Int, String)](
+      (CANON_EQ, "CANON_EQ(canonical equivalences)"),
+      (COMMENTS, "COMMENTS"),
+      (UNICODE_CASE, "UNICODE_CASE"),
+      (UNICODE_CHARACTER_CLASS, "UNICODE_CHARACTER_CLASS"),
+      (UNIX_LINES, "UNIX_LINES")
     )
-  }
 
-  def matches(regex: String, input: CharSequence): Boolean =
-    compile(regex).matcher(input).matches
+    for (i <- 0 until unsupportedOptions.length) {
+      notSupported(unsupportedOptions(i)._1, unsupportedOptions(i)._2)
+    }
 
-  def quote(s: String): String = {
-    s"\\Q${s}\\E"
+    // Any bit set other than given set throws.
+    if ((flags & ~(CASE_INSENSITIVE | DOTALL | LITERAL | MULTILINE)) != 0) {
+      throw new IllegalArgumentException(s"Unknown flag ${flags}")
+    }
   }
 
-  private object CompiledPatternStore {
-    final case class Key(regex: String, flags: Int)
-    final class Node(var key: Key,
-                     var value: RE2RegExpOps,
-                     var rc: Int,
-                     var next: Node)
-
-    private def freshNode(next: Node) =
-      new Node(null, new RE2RegExpOps(null), 0, next)
-
-    // The tip of Nodes. The Nodes form a ring buffer of some length.
-    private var last: Node = {
-      // Populate the ringbuffer
-      @tailrec def f(n: Node, num: Int): Node =
-        if (num < 0) {
-          n
-        } else {
-          f(freshNode(n), num - 1)
-        }
-      val last = freshNode(null)
-      last.next = f(last, 128)
-      last
-    }
+  private def toRe2Flags(flags: Int): Int = {
+
+    // Pass re2s only the flags it knows about and clear the rest.
+    //
+    // j.u.regex LITERAL is handled in j.u.regex.Pattern#compile
+    // so OK to clear that bit. java bits not known to re2s will cause
+    // it to throw, so clear them also.
+    //
+    // re2s supports only CASE_INSENSITIVE | DOTALL | MULTILINE |
+    //                    DISABLE_UNICODE_GROUPS))
+    //
+    // DISABLE_UNICODE_GROUPS causes rejectUnsupportedOptions()
+    // to be thrown, so only CASE_INSENSITIVE, DOTALL, and MULTILINE are
+    // left when execution reaches this point.
+    //
+    // The constants for these three definitely differ between j.u.regex
+    // and re2s and must be be translated.
+
+    val optionTranslations = Array[(Int, Int)](
+      (CASE_INSENSITIVE, re2s.Pattern.CASE_INSENSITIVE),
+      (DOTALL, re2s.Pattern.DOTALL),
+      (MULTILINE, re2s.Pattern.MULTILINE)
+    )
 
-    // Used to quickly look up a Node from a Key.
-    private val map = scala.collection.mutable.HashMap.empty[Key, Node]
-
-    private def selectNode(regex: String, flags: Int): Node = {
```

Comment:
The code here implements a bit overdesigned LRU cache. It would be good to keep something like that in the new implementation (i.e. cache compiled instances of regex per input string).

Reply:
I believe I removed that code when I converted to re2s.  

As much as I love caches, I convinced myself that the Pattern cache code 
had not much upside and a lot of down side.  

Much as I like caches, this seemed to me like one of those places to go simple and correct
until  real data shows the need.  Would a cache help at all?

To get the re2s transition to happen I decided to drop the code until experience
and data show that a cache is helpful.  The work could be done in a future sprint.

Compiling a pattern is basically running the parser over it and returning the same thing
if it parses. One must balance the cost of a fast parser versus a cache scheme with
memory allocation.  I made the bet that the former will be faster in almost all use cases.
Of course, there will be special cases, such as repeated use of the same pattern,
where i loose the bet.

Now, I am willing to believe that any serious user of regex is going to be caching themselves
and not relying upon the implementation. Rightly, the JVM documentation says nothing
about caching.

Your thoughts?

Reply:
A number of string methods internally use regex but don't allow user to cache the regular expressions such as `java.lang.String.{replaceAll, replaceFirst, split, matches}`. So caching the compilation of the regex can reduce overhead significantly, especially when performing those operations on small strings.

Reply:
Understood.  

Can this be spun off into its own Issue and not block this PR?  I think it is going to take
some days and some discussion to get regex pattern caching right.  For example,
informed gestimates of cache size (probably > 10 and < 100).  I must consider 
interactions with underlying RE2 caching. On a first run through it looks like straight
delegation will not work, but then again, tell an engineer that something will not work
and 48 hours later they will have a solution.

Lastly, I must make sure that whatever I write does not trigger issue #1091. Easy
enough to avoid as long as the defect is kept in mind.

Write an LRU cache seems like a job interview question, to be done in 5 minutes with
people judging:  guess I would fail that interview.  The simple list is obvious and
non-concurrent and I tend to write industrial strength code, perhaps at the expense
of timeliness.  That is my excuse for the day and I am sticking with it (Locale specific joke!)

Please advise.  Thank you.

Reply:
I would fail that interview too.

Reply:
It's perfectly fine to file this for later. Lets do that.